rubix-prestosql/src/main/java/com/qubole/rubix/prestosql/PrestoClusterManager.java

```diff
@@ -209,13 +56,21 @@ public boolean isMaster()
   @Override
   public List<String> getNodes()
   {
-    try {
-      return nodesCache.get("nodeList");
+    requireNonNull(nodeManager, "nodeManager is null");
+    List<Node> workers = nodeManager.getWorkerNodes().stream()
+            .filter(node -> !node.isCoordinator())
+            .collect(toImmutableList());
+
+    if (workers.isEmpty()) {
+      // Empty result set => server up and only master node running, return localhost has the only node
+      // Do not need to consider failed nodes list as 1node cluster and server is up since it replied to allNodesRequest
+      return ImmutableList.of(nodeManager.getCurrentNode().getHost());
     }
-    catch (ExecutionException e) {
-      log.warn("Error fetching node list : ", e);
-    }
-    return null;
+
+    return workers.stream()
+            .map(Node::getHost)
+            .sorted()
```

Comment:
`getNodes` is called frequently and parallely too. Though the list to sort here will be small but would be better to keep a cache of node list here. We can refactor the code to have the cache logic common to both the providers.

Reply:
> getNodes is called frequently and parallely too. Though the list to sort here will be small but would be better to keep a cache of node list here.

Makes sense.